### PR TITLE
feat(terraform): support more json encoded objects as part of terraform resource and fix evaluation of true/false in json

### PR DESCRIPTION
**Pull request by bo156**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sca|secrets|serverless|terraform|general|graph|terraform_plan
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

Changed `base_attribute_solver` to always try to load a json object. Up until now we only did it for `policy` attribute, but as many places could have this behavior I changed it to be the default.
I also added a new test that checks it using a custom policy for `aws_emr_cluster`

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my feature, policy, or fix is effective and works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py [hunk 1]
```diff
@@ -199,11 +199,27 @@ def _is_variable_dependant(value: Any, source: str) -> bool:
...
+        - resource "aws_emr_security_configuration" "example2" {
+            configuration = <<EOF
+                {
+                ... json-data
+                }
+            EOF
+        }
+        """
+        try:
+            value_to_check = json.loads(value_to_check)
```
Comment (gruebel) [inline checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py:221]: I don't think this is a good idea, because it will also change none JSON strings, like `"true" -> True` or `"12345" -> 12345`. We should at least do a string lookup, before trying to cast to dict/list.
Also not sure, if someone does it, but if someone actually wants the value to be a string, then we will now break it.
ex. `name = "[\"some name\"]"` so this wouldn't be a feature, it would be a breaking change.

**Review by gruebel — CHANGES_REQUESTED** [review]

File: checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py (hunk 1 above)
Comment (bo156) [reply checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py:221]: got you, my intention was we don't want to add support for each attribute terraform allow you to add as an inlne configuration (to reduce future maintenance cost).
Do you have any idea how to do it better?
If not, I will just regress it to the way it was before and add support for `configuration` as well

File: tests/terraform/graph/checks/resources/CustomAwsEMRSecurityConfiguration/main.tf [hunk 2]
```diff
@@ -0,0 +1,91 @@
+provider "aws" {
```
Comment (tronxd) [inline tests/terraform/graph/checks/resources/CustomAwsEMRSecurityConfiguration/main.tf:1]: Add a test of edge cases where the attribute is `"true"` (every boolean attribute), and "`5`" (you can use data retention attributes which accepts a number)

File: checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py (hunk 1 above)
Comment (tronxd) [inline checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py:221]: I'm worrying about performance degradation here, since we will try parsing every attribute into a JSON.
Implement a mechanism which lookups a JSON object starting delimiters (`{` for object , `[` for an array of objects), and parse into a JSON only when there's a match

Comment (bo156) [reply checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py:221]: Do you think it will be better to do it like that if you also relate to what @gruebel said about this part?
I mainly try to understand if it is even ok to try to manipulate each attribute like that, even if it does look like a json object.

Comment (gruebel) [reply checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py:221]: if we target a more general solution then this would be the minimum to do to make sure, we don't try to parse everything as json.
the other approach would be to maintain a map of resource + argument, which should be parsed as json.

File: tests/terraform/graph/checks/resources/CustomAwsEMRSecurityConfiguration/main.tf (hunk 2 above)
Comment (bo156) [reply tests/terraform/graph/checks/resources/CustomAwsEMRSecurityConfiguration/main.tf:1]: done

**Review by tronxd — APPROVED** [review]

File: checkov/terraform/checks/resource/aws/EMRClusterIsEncryptedKMS.py [hunk 3]
```diff
@@ -16,7 +16,7 @@ def scan_resource_conf(self, conf):
         if 'configuration' not in conf:
             return CheckResult.SKIPPED
         configuration = conf['configuration'][0]
-        if "SSE-KMS" in configuration:
+        if "SSE-KMS" in str(configuration):
```
Comment (gruebel) [inline checkov/terraform/checks/resource/aws/EMRClusterIsEncryptedKMS.py:19]: ok, this is needed, because it is now a `dict`? could change it to a `BaseValueResourceCheck`? then the inspected key should be something like `configuration. EncryptionConfiguration.AtRestEncryptionConfiguration. S3EncryptionConfiguration. EncryptionMode` and expected value `SSE-KMS`

File: checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py [hunk 4]
```diff
@@ -199,11 +199,29 @@ def _is_variable_dependant(value: Any, source: str) -> bool:
...
+            configuration = <<EOF
+                {
+                ... json-data
+                }
+            EOF
+        }
+        """
+        try:
+            # Check if the value looks like a json object, and if it is try to parse it
+            if isinstance(value_to_check, str) and value_to_check.startswith('{'):
```
Comment (gruebel) [inline checkov/common/checks_infra/solvers/attribute_solvers/base_attribute_solver.py:222]: let's go with that 👍

File: checkov/terraform/checks/resource/aws/EMRClusterIsEncryptedKMS.py (hunk 3 above)
Comment (bo156) [reply checkov/terraform/checks/resource/aws/EMRClusterIsEncryptedKMS.py:19]: Yeah it's beacuse it'ss a dict :)
I actually checked it out but I thought there might be more places to place `SSE-KMS` in (maybe not this field alone I mean). If there are no more places I will change it

**Review by gruebel — APPROVED** [review]
🍪
